rules/manager.go

```diff
@@ -212,21 +215,45 @@ func EngineQueryFunc(engine *promql.Engine, q storage.Queryable) QueryFunc {
 	}
 }
 
+func ExemplarQuerierQueryFunc(q storage.ExemplarQueryable) ExemplarQueryFunc {
+	return func(ctx context.Context, expr parser.Expr, ts time.Time, interval time.Duration) ([]exemplar.QueryResult, error) {
+		selectors := parser.ExtractSelectors(expr)
+		if len(selectors) < 1 {
+			return nil, fmt.Errorf("no selectors found on exemplar query")
+		}
+
+		eq, err := q.ExemplarQuerier(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		// Query all the raw exemplars that match the query
+		ex, err := eq.Select(timestamp.FromTime(ts.Add(-1*interval)), timestamp.FromTime(ts), selectors...)
+		if err != nil {
+			return nil, err
+		}
+
+		return ex, nil
+	}
+}
+
 // A Rule encapsulates a vector expression which is evaluated at a specified
 // interval and acted upon (currently either recorded or used for alerting).
 type Rule interface {
 	Name() string
 	// Labels of the rule.
 	Labels() labels.Labels
-	// eval evaluates the rule, including any associated recording or alerting actions.
+	// Eval evaluates the rule, including any associated recording or alerting actions.
 	Eval(context.Context, time.Time, QueryFunc, *url.URL, int) (promql.Vector, error)
+	// EvalWithExemplars evaluates the rule, including any associated recording or alerting actions and emit exemplars.
+	EvalWithExemplars(ctx context.Context, ts time.Time, interval time.Duration, query QueryFunc, exquery ExemplarQueryFunc, url *url.URL, limit int) (promql.Vector, []exemplar.QueryResult, error)
```

Comment:
I think it would be good to state the meaning of the parameters, at least `url`, `interval` and `limit`.
`exquery` is not Go-style naming; what about exemplarQuery ?

Reply:
i have updated the above to be consistent with `Eval`. if you can please give me an example of how to document each param, i would be happy to add in the comments.